src/platform/datapath_winkernel.c

```diff
@@ -1768,29 +1648,56 @@ QuicDataPathBindingCreate(
     return Status;
 }
 
-_IRQL_requires_max_(DISPATCH_LEVEL)
-void
-QuicDataPathBindingSocketClosed(
-    _In_ QUIC_DATAPATH_BINDING* Binding
-    )
-{
-    IoCleanupIrp(&Binding->SocketContext.Irp);
-    QuicRundownUninitialize(&Binding->ClientRundown);
-    QUIC_FREE(Binding);
-}
-
 _IRQL_requires_max_(PASSIVE_LEVEL)
 void
 QuicDataPathBindingDelete(
     _In_ QUIC_DATAPATH_BINDING* Binding
     )
 {
     QUIC_DBG_ASSERT(Binding != NULL);
-    //
-    // Close the socket (asynchronously) which will call the above function,
-    // QuicDataPathBindingSocketClosed, when it completes to finish.
-    //
-    QuicDataPathCleanupSocketContext(&Binding->SocketContext);
+    if (Binding->SocketContext.Socket != NULL) {
+
+        QUIC_EVENT CompletionEvent;
+        uint8_t IrpBuffer[sizeof(IRP) + sizeof(IO_STACK_LOCATION)];
+        PIRP Irp = (PIRP)IrpBuffer;
+
+        QuicEventInitialize(&CompletionEvent, FALSE, FALSE);
+        QuicZeroMemory(Irp, sizeof(IrpBuffer));
+        IoInitializeIrp(Irp, sizeof(IrpBuffer), 1);
+        IoSetCompletionRoutine(
+            Irp,
+            QuicDataPathIoCompletion,
+            &CompletionEvent,
+            TRUE,
+            TRUE,
+            TRUE);
+
+        NTSTATUS Status =
+            Binding->SocketContext.DgrmSocket->Dispatch->
+            WskCloseSocket(
+                Binding->SocketContext.Socket,
+                Irp);
+
+        if (Status == STATUS_PENDING) {
+            QuicEventWaitForever(CompletionEvent);
```

Comment:
@mtfriesen @nigriMSFT @maolson-msft I'm getting a hang here when running my tests:
```
3: kd> kn
  *** Stack trace for last set context - .thread/.cxr resets it
 # Child-SP          RetAddr           Call Site
00 fffffa85`d8837300 fffff805`5c82f09a nt!KiSwapContext+0x76 [minkernel\ntos\ke\amd64\ctxswap.asm @ 145] 
01 fffffa85`d8837440 fffff805`5c8308f3 nt!KiSwapThread+0x5ba [minkernel\ntos\ke\thredsup.c @ 11883] 
02 fffffa85`d8837500 fffff805`5c831119 nt!KiCommitThreadWait+0x153 [minkernel\ntos\ke\waitsup.c @ 795] 
03 fffffa85`d88375a0 fffff805`678b0ca0 nt!KeWaitForSingleObject+0x239 [minkernel\ntos\ke\wait.c @ 861] 
04 fffffa85`d8837690 fffff805`6785fd95 msquic!QuicDataPathBindingDelete+0x140 [F:\msquic\src\platform\datapath_winkernel.c @ 1683] 
05 fffffa85`d8837840 fffff805`678544df msquic!QuicBindingUninitialize+0x105 [F:\msquic\src\core\binding.c @ 202] 
06 fffffa85`d88378b0 fffff805`678694d8 msquic!QuicLibraryReleaseBinding+0x1af [F:\msquic\src\core\library.c @ 1326] 
07 fffffa85`d8837950 fffff805`6785893c msquic!QuicConnFree+0x398 [F:\msquic\src\core\connection.c @ 376] 
08 fffffa85`d88379d0 fffff805`67859ff8 msquic!QuicConnRelease+0x23c [F:\msquic\src\core\connection.h @ 934] 
09 fffffa85`d8837a30 fffff805`6785a0ca msquic!QuicWorkerProcessConnection+0x5f8 [F:\msquic\src\core\worker.c @ 539] 
0a fffffa85`d8837b10 fffff805`5c9432c5 msquic!QuicWorkerThread+0xba [F:\msquic\src\core\worker.c @ 571] 
0b fffffa85`d8837bf0 fffff805`5ca20ce8 nt!PspSystemThreadStartup+0x55 [minkernel\ntos\ps\psexec.c @ 9989] 
0c fffffa85`d8837c40 00000000`00000000 nt!KiStartSystemThread+0x28 [minkernel\ntos\ke\amd64\threadbg.asm @ 83] 
```
Any idea why or how to debug this? I have the kd attached.